clients/ts/signalr/src/ServerSentEventsTransport.ts

```diff
@@ -0,0 +1,109 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+import { DataReceived, TransportClosed } from "./Common";
+import { HttpClient } from "./HttpClient";
+import { IConnection } from "./IConnection";
+import { ILogger, LogLevel } from "./ILogger";
+import { ITransport, TransferFormat } from "./ITransport";
+import { Arg, getDataDetail, sendMessage } from "./Utils";
+
+export class ServerSentEventsTransport implements ITransport {
+    private readonly httpClient: HttpClient;
+    private readonly accessTokenFactory: () => string;
+    private readonly logger: ILogger;
+    private readonly logMessageContent: boolean;
+    private eventSource: EventSource;
+    private url: string;
+
+    constructor(httpClient: HttpClient, accessTokenFactory: () => string, logger: ILogger, logMessageContent: boolean) {
+        this.httpClient = httpClient;
+        this.accessTokenFactory = accessTokenFactory || (() => null);
+        this.logger = logger;
+        this.logMessageContent = logMessageContent;
+    }
+
+    public connect(url: string, transferFormat: TransferFormat, connection: IConnection): Promise<void> {
+        Arg.isRequired(url, "url");
+        Arg.isRequired(transferFormat, "transferFormat");
+        Arg.isIn(transferFormat, TransferFormat, "transferFormat");
+        Arg.isRequired(connection, "connection");
+
+        if (typeof (EventSource) === "undefined") {
+            throw new Error("'EventSource' is not supported in your environment.");
+        }
+
+        this.logger.log(LogLevel.Trace, "(SSE transport) Connecting");
+
+        this.url = url;
+        return new Promise<void>((resolve, reject) => {
+            if (transferFormat !== TransferFormat.Text) {
+                reject(new Error("The Server-Sent Events transport only supports the 'Text' transfer format"));
+            }
+
+            const token = this.accessTokenFactory();
+            if (token) {
+                url += (url.indexOf("?") < 0 ? "?" : "&") + `access_token=${encodeURIComponent(token)}`;
+            }
+
+            const eventSource = new EventSource(url, { withCredentials: true });
+
+            try {
+                eventSource.onmessage = (e: MessageEvent) => {
+                    if (this.onreceive) {
+                        try {
+                            this.logger.log(LogLevel.Trace, `(SSE transport) data received. ${getDataDetail(e.data, this.logMessageContent)}.`);
+                            this.onreceive(e.data);
+                        } catch (error) {
+                            if (this.onclose) {
+                                this.onclose(error);
+                            }
+                            return;
+                        }
+                    }
+                };
+
+                eventSource.onerror = (e: any) => {
+                    const error = new Error(e.message || "Error occurred");
+                    reject(error);
```

Comment:
Does order matter here? Reject at the end makes more sense to me

Reply:
Actually, we should do one or the other. If `onopen` has fired, we resolve the promise (to say we're connected), and we shouldn't then reject it later. If `onopen` hasn't fired, this is an error during the initial connection and we should reject the promise. I'll tidy this up as well.